### ISSUE TITLE
Czech keyboard

### ISSUE THREAD

**Issue by honzapulpan** [holzschu/a-shell#32]
Czech keyboard (probably others as well) has an option to write special characters with alt. For example alt+2 writes @. When I use external keyboard, these characters don’t work in a-shell. It looks like alt is used as standard shift key in this case.

**Comment by holzschu** [comment]
Interesting. I'll have a look.
I think it could be related to the "use caps-lock as escape" option. Is the problem still there when you deactivate the option?

**Comment by honzapulpan** [comment]
Yes, the problem still exists when I deactivate the option. 

Most of the special characters I use are at number keys. Alt+number writes number instead of special character. Except alt+1 and alt+5, which writes apostrophe. It is correct for alt+1, while alt+5 should write ~. And alt+9 writes [, while it should write {. Similar alt+0 writes ], while it should be }.

There is different behavior using iPython. Alt+1 writes apostrophe (correct), while alt+2,3 etc. writes nothing and down in status above keyboard bar “Repeat 2” (or 3...) is shown. 

It isn’t a huge problem though, I still can use English keyboard :-)

**Comment by holzschu** [comment]
This is interesting. Without "escape as caps-lock", we use iOS default keyboard management. I did notice it had issues in French with "ç/Ç" (alt-c), but I hoped other languages would be spared.

The good news is, since I already added something for the French keyboard, it should be easy to expand it for other keyboards. Is this the right layout? 
https://cdn.shopify.com/s/files/1/0810/3669/files/mac-czech-keyboard_1024x1024.png?12659750357617424768

I think the behaviour in iPython is related how it manages shortcuts, bypassing the system shortcuts.

**Comment by honzapulpan** [comment]
Yes, this layout is correct. Doesn't show all the ALT or SHIFT-ALT characters (most of them I never used :-)

Here are the full layouts with SHIFT and ALT keys (screens from MacOS, but I checked and it's the same at iPad)

CZ layout:
<img width="318" alt="cz_keyboard" src="https://user-images.githubusercontent.com/11267326/79120482-4d5fa600-7d93-11ea-98ed-fc488481736d.png">
with SHIFT:
<img width="318" alt="cz_keyboard_SHIFT" src="https://user-images.githubusercontent.com/11267326/79120485-4e90d300-7d93-11ea-948e-2069151ed92c.png">
with ALT:
<img width="318" alt="cz_keyboard_ALT" src="https://user-images.githubusercontent.com/11267326/79120483-4df83c80-7d93-11ea-9942-02c6acd09cc6.png">
with SHIFT+ALT:
<img width="318" alt="cz_keyboard_SHIFT_ALT" src="https://user-images.githubusercontent.com/11267326/79120486-4fc20000-7d93-11ea-8246-ea6d2294a007.png">

Thanks a lot. And it's really not a big issue, if you don't want to fix it :-) I can imagine it may be lot of work.

**Comment by holzschu** [comment]
I've added it to the TODO list: https://github.com/holzschu/a-shell/projects/1 
It's a long list, but I'll get to it.

**Comment by cristeahub** [comment]
One thing to note here is that in the newest iOS verison you can remap CAPS to ESC via the system settings for external keyboards. Maybe all the remapping done in the app can be removed.

The norwegian keyboard suffers greatly as it is right now. To type `{` one needs ALT+SHITFT+8. As it stands right now, that is not possible.

**Comment by holzschu** [comment]
The fix for Czech keyboards will be in Build 86, soon to be on TestFlight. During my tests, I found that left-alt and right-alt behaved differently, which was surprising.

@cristeahub do I understand correctly that there is a similar issue for Norwegian keyboards? On which keys, specifically? Now that I have specific treatment for 2 languages, adding a third is easier.

**Comment by cristeahub** [comment]
> @cristeahub do I understand correctly that there is a similar issue for Norwegian keyboards? On which keys, specifically? Now that I have specific treatment for 2 languages, adding a third is easier.

There are two issues:

- Within `vim` none of my option keys work. If I press option it seems to be ignored.

- When using the Norwegian layout some, but not all, of the alt modifiers are ignored. The number row seems to work fine, but the letters doesn't.

Here's the layout for reference:
No modifier keys:
<img width="341" alt="image" src="https://user-images.githubusercontent.com/1488921/82951972-71c0bc00-9fa8-11ea-8709-f9c00f3bffcd.png">

With shift:
<img width="339" alt="image" src="https://user-images.githubusercontent.com/1488921/82952179-c95f2780-9fa8-11ea-8074-79794dbf16db.png">

With option:
<img width="320" alt="image" src="https://user-images.githubusercontent.com/1488921/82952209-d845da00-9fa8-11ea-955c-7cd690c60659.png">

With option+shift:
<img width="338" alt="image" src="https://user-images.githubusercontent.com/1488921/82952233-e3006f00-9fa8-11ea-9e4c-341e4cceda07.png">

**Comment by honzapulpan** [comment]
Thank you @holzschu. I tested it at build 86 and Czech keyboard is working perfectly fine.

**Comment by holzschu** [comment]
@cristeahub: about Vim and alt-keys, at least on my keyboard, left-alt and right-alt behave differently: left-alt-c sends escape-C, right-alt-C sends "ç". Which is extremely weird, because I didn't think there was a difference between them. I'm looking into it. That is also the reason why alt-keys don't work.

Support for Norwegian keyboards is going to be in the next build (93).

**Comment by holzschu** [comment]
Hi, 
build 94 has a generic fix (alt-key should now work with all external keyboards). I removed the code specific to Czech, English or Norwegian keyboards. Please let me know if this doesn't work.

**Comment by honzapulpan** [comment]
Czech one works perfectly fine. Now also in ipython. Thanks!

**Comment by cristeahub** [comment]
This works great! Thanks a lot! :+1:
